### PR TITLE
feat(favorite): #MAG-219 create card favorite endpoint

### PR DESCRIPTION
**Pull request by Alex-Delaveau**
## Describe your changes
I have created a new endpoint @Put("/card/favorite/:id") that allows changing the favorite status of a card for a user.

The card's id is included in the parameters, and the body of the request contains a boolean:

- true: it removes the user from the favorites list of this card.

- false: it adds the user to the favorites list of this card.

In addition, I have introduced a system to trigger custom exceptions that are used to return a specific status code to the user in case of a problem in the program.
## Checklist tests
This can be tested with Postman by sending requests to the endpoint @Put("/card/favorite/:id"). If the boolean is true and you're not on the favorite list, or if the boolean is false and you're already on the favorite list, an HTTP error 400 (Bad Request) will be returned. If the card id you provided does not exist, you will receive an HTTP 404 (Not Found) error.

Alternatively, you can test it with [MAG-221](https://jira.support-ent.fr/browse/MAG-221)
## Issue ticket number and link
[MAG-219](https://jira.support-ent.fr/browse/MAG-219)
## Checklist before requesting a review (magic string, indentation, comment/documentation...)

- [x] I have detailed the tests to do in my feature/fix in order to prevent consequents regressions (must specify in **Checklist tests**)
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (API Doc etc...) - (must specify in **Description** for target version)
- [ ] If it is a consequent feature, I have added thorough tests.
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to this project (must specify in **Description**)

### CODE REVIEW

File: src/main/java/fr/cgi/magneto/controller/CardController.java [hunk 1]
```diff
@@ -185,6 +187,23 @@ public void update(HttpServerRequest request) {
             });
         });
     }
+    @Put("/card/favorite/:id")
```
Comment (flmartineau) [inline src/main/java/fr/cgi/magneto/controller/CardController.java:190]: `@Put("/card/:id/favorite")` correspondrait plus à la convention RESTful (l'id suit l'objet qui lui correspond, ici une "card".

**Review by flmartineau — APPROVED** [review]

File: src/main/java/fr/cgi/magneto/controller/CardController.java [hunk 2]
```diff
@@ -185,6 +187,23 @@ public void update(HttpServerRequest request) {
...
         });
     }
+    @Put("/card/:id/favorite")
+    @ApiDoc("Update the favorites of a card")
+    @ResourceFilter(ReadBoardRight.class)
+    @SecuredAction(value = "", type = ActionType.RESOURCE)
+    @Trace(Actions.FAVORITE_UPDATE)
+    public void updateFavorite(HttpServerRequest request){
+        RequestUtils.bodyToJson(request, pathPrefix + "cardUpdateFavorite", body -> {
+            UserUtils.getUserInfos(eb, request, user -> {
```
Comment (qperie) [inline src/main/java/fr/cgi/magneto/controller/CardController.java:197]: Cas où user == null

File: src/main/java/fr/cgi/magneto/controller/CardController.java [hunk 3]
```diff
@@ -185,6 +187,23 @@ public void update(HttpServerRequest request) {
...
+                String cardId = request.getParam(Field.ID);
+                boolean favorite = body.getBoolean(Field.IS_FAVORITE);
+                cardService.updateFavorite(cardId, favorite, user.getUserId())
+                        .onSuccess(res -> renderJson(request, res))
+                        .onFailure(err -> HttpRequestHelper.sendError(request, err));
```
Comment (qperie) [inline src/main/java/fr/cgi/magneto/controller/CardController.java:202]: log.error()

File: src/main/java/fr/cgi/magneto/security/ReadBoardRight.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+            if (boardId == null || boardId.isEmpty()) {
+                getBoardIdByCardId(cardId, mongo)
+                        .onSuccess(boardIdRes -> {
+                            if (boardIdRes == null) {
+                                HttpRequestHelper.sendError(request, new RessourceNotFoundException("Card not found"));
+                            }
+                            handleHasReadBoardRight(request, user, boardIdRes, handler);
+                        })
+                        .onFailure(err -> {
+                            HttpRequestHelper.sendError(request, err);
```
Comment (qperie) [inline src/main/java/fr/cgi/magneto/security/ReadBoardRight.java:50]: log.error()

File: src/main/java/fr/cgi/magneto/security/ReadBoardRight.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+                .put(Field._ID, boardId)
+                .put(Field.DELETED, false)
+                .put(Mongo.OR,
+                        new JsonArray()
+                                .add(new JsonObject()
+                                        .put(Field.OWNERID, user.getUserId()))
+                                .add(new JsonObject()
+                                        .put(String.format("%s.%s", Field.SHARED, Field.USERID),
+                                                new JsonObject().put(Mongo.IN, new JsonArray().add(user.getUserId())))
+                                        .put(String.format("%s.%s", Field.SHARED, "fr-cgi-magneto-controller-ShareBoardController|initReadRight"), true))
```
Comment (qperie) [inline src/main/java/fr/cgi/magneto/security/ReadBoardRight.java:75]: Magic string

**Pull request merged**